File: generate_images.py
#!/usr/bin/python3

import asyncio
import os
import re

import aiohttp

from github_stats import Stats


################################################################################
# Helper Functions
################################################################################

def generate_output_folder() -> None:
    """
    Create the output folder if it does not already exist
    """
    if not os.path.isdir("generated"):
        os.mkdir("generated")


################################################################################
# Individual Image Generation Functions
################################################################################

async def generate_overview(s: Stats) -> None:
    """
    Generate an SVG badge with summary statistics
    :param s: Represents user's GitHub statistics
    """
    try:
        print("Generating overview SVG...")
        
        with open("templates/overview.svg", "r") as f:
            output = f.read()

        # Get stats with error handling
        name = await s.name
        stars = await s.stargazers
        forks = await s.forks
        contributions = await s.total_contributions
        lines_changed = await s.lines_changed
        views = await s.views
        repos = await s.all_repos
        
        # Replace template variables
        output = re.sub("{{ name }}", name, output)
        output = re.sub("{{ stars }}", f"{stars:,}", output)
        output = re.sub("{{ forks }}", f"{forks:,}", output)
        output = re.sub("{{ contributions }}", f"{contributions:,}", output)
        
        changed = lines_changed[0] + lines_changed[1]
        output = re.sub("{{ lines_changed }}", f"{changed:,}", output)
        output = re.sub("{{ views }}", f"{views:,}", output)
        output = re.sub("{{ repos }}", f"{len(repos):,}", output)

        generate_output_folder()
        with open("generated/overview.svg", "w") as f:
            f.write(output)
        
        print("Overview SVG generated successfully!")
        
    except Exception as e:
        print(f"Error generating overview SVG: {e}")
        raise


async def generate_languages(s: Stats) -> None:
    """
    Generate an SVG badge with summary languages used
    :param s: Represents user's GitHub statistics
    """
    try:
        print("Generating languages SVG...")
        
        with open("templates/languages.svg", "r") as f:
            output = f.read()

        progress = ""
        lang_list = ""
        
        # Get languages with error handling
        languages = await s.languages
        sorted_languages = sorted(languages.items(), reverse=True,
                                  key=lambda t: t[1].get("size"))
        
        delay_between = 150
        for i, (lang, data) in enumerate(sorted_languages):
            color = data.get("color")
            color = color if color is not None else "#000000"
            ratio = [.98, .02]
            if data.get("prop", 0) > 50:
                ratio = [.99, .01]
            if i == len(sorted_languages) - 1:
                ratio = [1, 0]
            progress += (f'<span style="background-color: {color};'
                         f'width: {(ratio[0] * data.get("prop", 0)):0.3f}%;'
                         f'margin-right: {(ratio[1] * data.get("prop", 0)):0.3f}%;" '
                         f'class="progress-item"></span>')
            lang_list += f"""
<li style="animation-delay: {i * delay_between}ms;">
<svg xmlns="http://www.w3.org/2000/svg" class="octicon" style="fill:{color};"
viewBox="0 0 16 16" version="1.1" width="16" height="16"><path
fill-rule="evenodd" d="M8 4a4 4 0 100 8 4 4 0 000-8z"></path></svg>
<span class="lang">{lang}</span>
<span class="percent">{data.get("prop", 0):0.2f}%</span>
</li>

"""

        output = re.sub(r"{{ progress }}", progress, output)
        output = re.sub(r"{{ lang_list }}", lang_list, output)

        generate_output_folder()
        with open("generated/languages.svg", "w") as f:
            f.write(output)
        
        print("Languages SVG generated successfully!")
        
    except Exception as e:
        print(f"Error generating languages SVG: {e}")
        raise


################################################################################
# Main Function
################################################################################

async def main() -> None:
    """
    Generate all badges
    """
    # Validate required environment variables
    access_token = os.getenv("ACCESS_TOKEN")
    if not access_token:
        # Try fallback to GITHUB_TOKEN as a last resort
        access_token = os.getenv("GITHUB_TOKEN")
        if not access_token:
            raise Exception("ACCESS_TOKEN environment variable is required! Please set it in your repository secrets.")
        else:
            print("Warning: Using GITHUB_TOKEN as fallback. For better functionality, use a personal access token as ACCESS_TOKEN.")
    
    user = os.getenv("GITHUB_ACTOR")
    if not user:
        raise Exception("GITHUB_ACTOR environment variable is required!")
    
    # Validate and process optional environment variables
    exclude_repos = os.getenv("EXCLUDED", "").strip()
    exclude_repos = ({x.strip() for x in exclude_repos.split(",") if x.strip()}
                     if exclude_repos else None)
    
    exclude_langs = os.getenv("EXCLUDED_LANGS", "").strip()
    exclude_langs = ({x.strip() for x in exclude_langs.split(",") if x.strip()}
                     if exclude_langs else None)
    
    # Handle COUNT_STATS_FROM_FORKS more safely
    count_forks_env = os.getenv("COUNT_STATS_FROM_FORKS", "").strip()
    consider_forked_repos = count_forks_env.lower() in ("true", "1", "yes", "on") if count_forks_env else False
    
    print(f"Generating stats for user: {user}")
    print(f"Consider forked repositories: {consider_forked_repos}")
    if exclude_repos:
        print(f"Excluded repositories: {exclude_repos}")
    if exclude_langs:
        print(f"Excluded languages: {exclude_langs}")
    
    try:
        async with aiohttp.ClientSession() as session:
            s = Stats(user, access_token, session, exclude_repos=exclude_repos,
                      exclude_langs=exclude_langs,
                      consider_forked_repos=consider_forked_repos)
            await asyncio.gather(generate_languages(s), generate_overview(s))
        print("Successfully generated all stats images!")
    except Exception as e:
        print(f"Error generating stats: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: generate_images.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/github_stats.php';

function generateOutputFolder(): void
{
    if (!is_dir(__DIR__ . '/generated')) {
        mkdir(__DIR__ . '/generated', 0775, true);
        echo "Created generated/ directory\n";
    }
}

function validateTemplates(): void
{
    $requiredTemplates = [
        __DIR__ . '/templates/overview.svg' => ['{{ name }}', '{{ stars }}', '{{ forks }}', '{{ contributions }}', '{{ lines_changed }}', '{{ views }}', '{{ repos }}'],
        __DIR__ . '/templates/languages.svg' => ['{{ progress }}', '{{ lang_list }}'],
    ];

    foreach ($requiredTemplates as $templatePath => $requiredPlaceholders) {
        if (!is_file($templatePath)) {
            throw new RuntimeException("Template file {$templatePath} not found!");
        }

        $content = (string) file_get_contents($templatePath);
        foreach ($requiredPlaceholders as $placeholder) {
            if (!str_contains($content, $placeholder)) {
                throw new RuntimeException("Template {$templatePath} is missing placeholder: {$placeholder}");
            }
        }
    }

    echo "All template files validated successfully!\n";
}

function generateOverview(Stats $stats): void
{
    echo "Generating overview SVG...\n";
    $templatePath = __DIR__ . '/templates/overview.svg';
    if (!is_file($templatePath)) {
        throw new RuntimeException('Template file templates/overview.svg not found!');
    }

    $output = (string) file_get_contents($templatePath);
    $name = $stats->getName();
    $stargazers = $stats->getStargazers();
    $forks = $stats->getForks();
    $contributions = $stats->getTotalContributions();
    [$added, $deleted] = $stats->getLinesChanged();
    $views = $stats->getViews();
    $repos = $stats->getAllRepos();

    echo sprintf(
        "Stats collected - Name: %s, Stars: %s, Forks: %s, Contributions: %s\n",
        $name,
        number_format($stargazers),
        number_format($forks),
        number_format($contributions)
    );

    $replaceMap = [
        '{{ name }}' => $name,
        '{{ stars }}' => number_format($stargazers),
        '{{ forks }}' => number_format($forks),
        '{{ contributions }}' => number_format($contributions),
        '{{ lines_changed }}' => number_format($added + $deleted),
        '{{ views }}' => number_format($views),
        '{{ repos }}' => number_format(count($repos)),
    ];

    $output = str_replace(array_keys($replaceMap), array_values($replaceMap), $output);
    generateOutputFolder();

    $outputPath = __DIR__ . '/generated/overview.svg';
    file_put_contents($outputPath, $output);

    if (!is_file($outputPath)) {
        throw new RuntimeException('Failed to create overview.svg file');
    }
    $fileSize = filesize($outputPath);
    if ($fileSize === false || $fileSize < 100) {
        throw new RuntimeException('Generated overview.svg seems too small');
    }

    echo "Overview SVG generated successfully! ({$fileSize} bytes)\n";
}

function generateLanguages(Stats $stats): void
{
    echo "Generating languages SVG...\n";
    $templatePath = __DIR__ . '/templates/languages.svg';
    if (!is_file($templatePath)) {
        throw new RuntimeException('Template file templates/languages.svg not found!');
    }

    $output = (string) file_get_contents($templatePath);
    $languages = $stats->getLanguages();
    if ($languages === []) {
        echo "Warning: No languages found in repositories\n";
        $languages = [
            'Unknown' => [
                'size' => 1,
                'color' => '#cccccc',
                'prop' => 100.0,
            ],
        ];
    }

    uasort(
        $languages,
        static fn(array $a, array $b): int => ((int) ($b['size'] ?? 0)) <=> ((int) ($a['size'] ?? 0))
    );

    echo 'Found ' . count($languages) . " languages\n";

    $progress = '';
    $langList = '';
    $delayBetween = 150;
    $index = 0;
    $lastIndex = count($languages) - 1;
    foreach ($languages as $lang => $data) {
        $color = (string) ($data['color'] ?? '#000000');
        if ($color === '') {
            $color = '#000000';
        }
        $prop = (float) ($data['prop'] ?? 0.0);
        $ratio = [0.98, 0.02];
        if ($prop > 50) {
            $ratio = [0.99, 0.01];
        }
        if ($index === $lastIndex) {
            $ratio = [1.0, 0.0];
        }
        $progress .= sprintf(
            '<span style="background-color: %s;width: %.3f%%;margin-right: %.3f%%;" class="progress-item"></span>',
            $color,
            $ratio[0] * $prop,
            $ratio[1] * $prop
        );

        $langList .= sprintf(
            "\n<li style=\"animation-delay: %dms;\">\n<svg xmlns=\"http://www.w3.org/2000/svg\" class=\"octicon\" style=\"fill:%s;\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\"><path fill-rule=\"evenodd\" d=\"M8 4a4 4 0 100 8 4 4 0 000-8z\"></path></svg>\n<span class=\"lang\">%s</span>\n<span class=\"percent\">%.2f%%</span>\n</li>\n",
            $index * $delayBetween,
            $color,
            htmlspecialchars((string) $lang, ENT_QUOTES | ENT_XML1),
            $prop
        );

        $index++;
    }

    $output = str_replace(
        ['{{ progress }}', '{{ lang_list }}'],
        [$progress, $langList],
        $output
    );

    generateOutputFolder();
    $outputPath = __DIR__ . '/generated/languages.svg';
    file_put_contents($outputPath, $output);

    if (!is_file($outputPath)) {
        throw new RuntimeException('Failed to create languages.svg file');
    }
    $fileSize = filesize($outputPath);
    if ($fileSize === false || $fileSize < 100) {
        throw new RuntimeException('Generated languages.svg seems too small');
    }

    echo "Languages SVG generated successfully! ({$fileSize} bytes)\n";
}

function parseSetFromEnv(string $key): array
{
    $value = trim((string) getenv($key));
    if ($value === '') {
        return [];
    }
    $parts = array_map('trim', explode(',', $value));
    return array_values(array_filter($parts, static fn(string $item): bool => $item !== ''));
}

function envEnabled(string $key): bool
{
    $value = strtolower(trim((string) getenv($key)));
    if ($value === '') {
        return false;
    }
    return !in_array($value, ['0', 'false', 'no', 'off'], true);
}

function main(): int
{
    try {
        echo "Starting GitHub Stats generation...\n";
        validateTemplates();

        $accessToken = trim((string) getenv('ACCESS_TOKEN'));
        if ($accessToken === '') {
            $accessToken = trim((string) getenv('GITHUB_TOKEN'));
            if ($accessToken === '') {
                throw new RuntimeException('ACCESS_TOKEN environment variable is required! Please set it in your repository secrets.');
            }
            echo "Warning: Using GITHUB_TOKEN as fallback. For better functionality, use a personal access token as ACCESS_TOKEN.\n";
        }

        $user = trim((string) getenv('GITHUB_ACTOR'));
        if ($user === '') {
            throw new RuntimeException('GITHUB_ACTOR environment variable is required!');
        }

        $excludeRepos = parseSetFromEnv('EXCLUDED');
        $excludeLangs = parseSetFromEnv('EXCLUDED_LANGS');
        $considerForkedRepos = envEnabled('COUNT_STATS_FROM_FORKS');

        echo "Configuration:\n";
        echo "  User: {$user}\n";
        echo '  Consider forked repositories: ' . ($considerForkedRepos ? 'True' : 'False') . "\n";
        if ($excludeRepos !== []) {
            echo '  Excluded repositories: ' . implode(', ', $excludeRepos) . "\n";
        }
        if ($excludeLangs !== []) {
            echo '  Excluded languages: ' . implode(', ', $excludeLangs) . "\n";
        }

        $stats = new Stats($user, $accessToken, $excludeRepos, $excludeLangs, $considerForkedRepos);

        generateLanguages($stats);
        generateOverview($stats);
        echo "Successfully generated all stats images!\n";
        return 0;
    } catch (Throwable $e) {
        fwrite(STDERR, 'Error generating stats: ' . $e->getMessage() . "\n");
        return 1;
    }
}

exit(main());

